src/minio/core/policy_creator.py

```diff
@@ -0,0 +1,147 @@
+    #
+    # Creates policy with access to s3a://cdm-task-service/task-service/
+    # "task_service": {
+    #     "bucket": "cdm-task-service",
+    #     "base_prefix": "task-service",
+    #     "user_scoped": False,
+    #     "permission_level": PolicyPermissionLevel.READ,
+    # },
+    #
+    # Creates policy with access to s3a://cdm-user-scratch/
+    # "scratch": {
+    #     "bucket": "cdm-user-scratch",
+    #     "base_prefix": "",
+    #     "user_scoped": False,
+    #     "permission_level": PolicyPermissionLevel.WRITE,
+    # },
+}
+
+
+class PolicyCreator:
+    """
+    Creator for building new MinIO policies from scratch.
+    """
+
+    def __init__(
+        self,
+        policy_type: PolicyType,
+        target_name: str,
+        config: MinIOConfig,
+    ):
+        """
+        Initialize the PolicyCreator for building a new policy from scratch.
+
+        Args:
+            policy_type: Type of policy being created (USER_HOME, USER_SYSTEM, GROUP_HOME)
+            target_name: Name of the target (username or group name)
+            config: MinIO configuration for bucket and path information
+        """
+
+        self.policy_type = policy_type
+        self.target_name = target_name
+        self.system_config = SYSTEM_RESOURCE_CONFIG
+
+        self.config = config
+        # user warehouse for spark tables
+        self.user_sql_warehouse_path = f"s3a://{self.config.default_bucket}/{self.config.users_sql_warehouse_prefix}/{self.target_name}"
+        # user warehouse for general files
+        self.user_general_warehouse_path = f"s3a://{self.config.default_bucket}/{self.config.users_general_warehouse_prefix}/{self.target_name}"
+        # group warehouse for general files
+        self.group_general_warehouse_path = f"s3a://{self.config.default_bucket}/{self.config.groups_general_warehouse_prefix}/{self.target_name}"
+
+        # Internal section management with Dict[PolicySectionType, List[PolicyStatement]]
+        self._sections: Dict[PolicySectionType, List[PolicyStatement]] = {
+            PolicySectionType.GLOBAL_PERMISSIONS: [],
+            PolicySectionType.BUCKET_ACCESS: [],
+            PolicySectionType.READ_PERMISSIONS: [],
+            PolicySectionType.WRITE_PERMISSIONS: [],
+            PolicySectionType.DELETE_PERMISSIONS: [],
+        }
+
+    def _get_user_system_paths(
+        self, username: str
+    ) -> Dict[str, List[tuple[str, PolicyPermissionLevel]]]:
+        """Get system resource paths and permission levels for a user using the global configuration."""
+        user_paths = {}
+
+        for _, resource_config in self.system_config.items():
+            bucket = resource_config["bucket"]
+            base_prefix = resource_config["base_prefix"]
+            user_scoped = resource_config.get("user_scoped", True)
+            permission_level = resource_config.get(
+                "permission_level", PolicyPermissionLevel.WRITE
+            )
```

Comment:
If this is an internal implementation I would just make these fields required and therefore unambiguous

Reply:
:+1: